### PR TITLE
Implement a new error strategy for simplifying error messages

### PR DESCRIPTION
**Pull request by jimidle**
Here we implement a new error strategy for constructing error messages such that they are actually useful instead of listing out 780 possible tokens that were expected.

### CODE REVIEW

File: core/src/main/scala/com/databricks/labs/remorph/parsers/SqlErrorStrategy.scala [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  }
+  // $COVERAGE-ON$
+
+  override protected def reportInputMismatch(recognizer: Parser, e: InputMismatchException): Unit = {
+    val msg = new StringBuilder()
+    msg.append(getTokenErrorDisplay(e.getOffendingToken))
+    msg.append(" was unexpected ")
+    msg.append(generateMessage(recognizer, e))
+    msg.append("\nexpecting one of: ")
+    msg.append(buildExpectedMessage(recognizer, e.getExpectedTokens))
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/parsers/SqlErrorStrategy.scala:41]: use interpolated stings: `val msg = s"${getTokenErrorDisplay(e.getOffendingToken)}  was unexpected ..."` - more readable

File: core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlErrorStrategy.scala [hunk 2]
```diff
@@ -0,0 +1,963 @@
...
+
+    // We build the messages by looking at the stack trace of the exception, but if the
+    // rule translation is not found, or it is the same as the previous message, we skip it,
+    // to avoid repeating the same message multiple times. This is because a recognition error
+    // could be found in a parent rule or a child rule but there is no extra information
+    // provided by being more specific about the rule location. ALos, in some productions
+    // we may be embedded very deeply in the stack trace, so we want to avoid too many contexts
+    // in a message.
+    val stack = e.getStackTrace
+    stack.foreach { traceElement =>
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlErrorStrategy.scala:47]:
```suggestion
    val messages = e.getStackTrace.flatMap { traceElement => ...
```
flatMap would be more idiomatic here.

File: core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlErrorStrategy.scala [hunk 3]
```diff
@@ -0,0 +1,963 @@
...
+  /**
+   * When building the list of expected tokens, we do some custom manipulation so that we do not produce a list of 750
+   * possible tokens because there are so many keywords that can be used as id/column names. If ID is a valid expected
+   * token, then we remove all the keywords that are there because they can be an ID.
+   * @param expected
+   *   the set of valid tokens at this point in the parse, where the error was found
+   * @return
+   *   the expected string with tokens renamed in more human friendly form
+   */
+  import scala.collection.mutable
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlErrorStrategy.scala:82]: can you move the import to the top?

**Review by nfx — APPROVED** [review]
this definitely makes errors more digestible. few nits to make it a bit more idiomatic.

File: core/src/main/scala/com/databricks/labs/remorph/parsers/SqlErrorStrategy.scala (hunk 1 above)
Comment (jimidle) [reply core/src/main/scala/com/databricks/labs/remorph/parsers/SqlErrorStrategy.scala:41]: Will. do - @vil1 also suggested the same

**Pull request merged**